### DISCUSSION TITLE
support for timeseries

### DISCUSSION THREAD

**Discussion by gde-2** [electric-sheep-co/arduino-redis#44] [Ideas]
Hi all,

As I needed TimeSeries support from a Arduino Redis library I updated the redis.h and redis.cpp to support the redis TSADD operation. I would like to share my changes so thatit can be reviewed and if approved added to the distribution. Any proposal how to do this?

Regarding the code here are the changes:
redis.h:
bool tsadd(const char *key, unsigned long timestamp, const int value);

redis.cpp:
bool Redis::tsadd(const char *key, unsigned long timestamp, const int value)
{
    TRCMD_EXPECTOK("TS.ADD" , key, String(timestamp)+"000", String(value));
}

**Comment by rpj** [comment]
Hi @gde-2,

Thanks so much for getting in touch about the changes you made to support the TSADD operation, that is fantastic! I'd be more than happy to incorporate your changes into the library itself, and fortunately it's a pretty simple process as well.

Simply open a [pull request](https://github.com/electric-sheep-co/arduino-redis/discussions/44) with these changes, assign it to me & I'll merge them into the main branch and produce a new release! 

If you're unfamiliar with pull requests I'd be happy to answer any questions, but hopefully these great docs from Github themselves will provide some great information:
  * [_About pull requests_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests)
  * [_Creating a pull request_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request)
  * [_Collaborating with pull requests_](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests) -- very comprehensive documentation on the entire process

Thanks again for reaching out and I really look forward to helping you get your changes merged, I'm certain the community will appreciate them as well! 🙌 🚀

**Reply by rpj** [reply]
@gde-2 I went ahead and created issue #45 to properly track your pull request 👍

**Comment by ErMaVi** [comment]
Hi, thanks for the timeseries support! Could we also add the following: this will use whatever timestamp is in the server so makes it easier to add an entry to the series when you don't have a time-synced micro-controller:

diff --git a/Redis.cpp b/Redis.cpp
index d181d48..4a58ee7 100644
--- a/Redis.cpp
+++ b/Redis.cpp
@@ -151,6 +151,12 @@ bool Redis::tsadd(const char *key, unsigned long timestamp, const int value)
 {
     TRCMD_EXPECTOK("TS.ADD" , key, String(timestamp)+"000", String(value));
 }
+
+bool Redis::tsaddsrvts(const char *key, const float value)
+{
+    TRCMD_EXPECTOK("TS.ADD" , key, "*", String(value));
+}
+

 String Redis::info(const char *section)
 {
diff --git a/Redis.h b/Redis.h
index 434c954..7f7416c 100644
--- a/Redis.h
+++ b/Redis.h
@@ -359,6 +359,11 @@ public:
   bool unsubscribe(const char *channelOrPattern);

   bool tsadd(const char *key, unsigned long timestamp, const int value);
+
+  /*
+   * Adds key,value to redis time series using the server timestamp
+   */
+  bool tsaddsrvts(const char *key, const float value);

**Reply by rpj** [reply]
Hi @ErMaVi,

Sorry for the shuffle in comments here, I needed to reorganize.

I'm considering how best to approach this change because I don't particularly like the idea of adding a command to the Redis class that doesn't exist in the proper Redis command set: as you've probably noticed, all the other methods on Redis are named precisely to match their equivalent Redis command (sans the 'dot' character of course).

I've filed https://github.com/electric-sheep-co/arduino-redis/issues/48 and am thinking today on how best to approach implementing this. I will have an update to the library with this support very soon, thanks a ton for your eagerness & interest!

**Reply by rpj** [reply]
@ErMaVi I've just published [v2.4.4](https://github.com/electric-sheep-co/arduino-redis/releases/tag/2.4.4) with this support! I modified it a bit: simply pass a negative value (I'd suggest just `-1`) to `tsadd` and it will now use automatic server timestamps instead. It'll take a few hours to show up in the Arduino Library Manager. Thank you!

**Reply by ErMaVi** [reply]
Thank you!
